azurelinuxagent/common/protocol/extensions_goal_state.py

```diff
@@ -445,3 +493,32 @@ def __parse_and_add_extension_settings(settings_node, name, ext_handler, depends
             ext.certificateThumbprint = thumbprint
             ext_handler.properties.extensions.append(ext)
 
+
+class _ExtensionsGoalStateFromVmSettings(ExtensionsGoalState):
+    def __init__(self, etag, json_text):
+        super(_ExtensionsGoalStateFromVmSettings, self).__init__()
+        self._id = etag
+        self._text = json_text
+        self._type = 'VmSettings'
+
+        try:
+            self._parse_vm_settings(json_text)
+        except Exception as e:
+            # TODO: Review code to determine whether we can rename ExtensionConfigError to InvalidExtensionsGoalStateError or
+            #       if we need to differentiate between errors in ExtensionsConfig and vmSettings.
+            raise ExtensionConfigError("Error parsing vmSettings (etag: {0}): {1}\n{2}".format(etag, format_exception(e), json_text))
+
+    def _parse_vm_settings(self, json_text):
+        vm_settings = json.loads(json_text)
+        self._parse_required_features(vm_settings)
+        # TODO: Parse all atttributes
+
+    def _parse_required_features(self, vm_settings):
+        required_features = vm_settings.get("requiredFeatures")
+        if required_features is not None:
+            if not isinstance(required_features, list):
+                raise Exception("requiredFeatures should be an array")
+            self.required_features.append(required_features)
```

Comment:
Shouldn't this be `extend` over `append`?

Reply:
good catch, will fix it

Reply:
done